core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1761,7 +1822,7 @@ private[spark] class DAGScheduler(
             if (runningStages.contains(shuffleStage) && shuffleStage.pendingPartitions.isEmpty) {
               if (!shuffleStage.shuffleDep.shuffleMergeFinalized &&
                 shuffleStage.shuffleDep.getMergerLocs.nonEmpty) {
```

Comment:
Shall we add the condition `getMergerLocs.nonEmpty` into `shuffleDep.shuffleMergeFinalized`? e.g.,

```scala
  def shuffleMergeFinalized: Boolean = {
    // Empty RDD won't be computed therefore shuffle merge finalized should be true by default.
    if (shuffleMergeEnabled && numPartitions > 0 && mergerLocs.nonEmpty) {
      _shuffleMergedFinalized
    } else {
      true
    }
  }
```

Reply:
`mergerLocs` would be initially empty right when we check [here](https://github.com/apache/spark/blob/5880df41f50210f2df44f25640437d99b8979d70/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1402) to `prepareShuffleServicesForShuffleMapStage`. I don't think we can add `mergerLocs.nonEmpty` in `shuffleMergeFinalized`.